onnxruntime/core/providers/cpu/ml/linearregressor.cc

```diff
@@ -1,61 +1,59 @@
-  ORT_ENFORCE(info.GetAttrs<float>("coefficients", coefficients_).IsOK());
+static T GetAttr(const OpKernelInfo& info, const std::string& name) {
+  T value;
+  ORT_ENFORCE(info.GetAttr<T>(name, &value).IsOK());
+  return value;
 }
 
-template <>
-Status LinearRegressor<float>::Compute(OpKernelContext* ctx) const {
-  const auto* X = ctx->Input<Tensor>(0);
-  if (X->Shape().NumDimensions() == 0) {
-    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
-                  "Input shape needs to be at least a single dimension.");
-  }
-
-  int64_t stride = X->Shape().NumDimensions() == 1 ? X->Shape()[0] : X->Shape()[1];
-  int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
-  Tensor* Y = ctx->Output(0, TensorShape({N, targets_}));
-  const auto* Xdata = X->template Data<float>();
-  int64_t yindex = 0;
+template <typename T>
+static Eigen::Tensor<T, 2, Eigen::RowMajor, Eigen::DenseIndex> GetCoefficients(const OpKernelInfo& info, int64_t targets_) {
+  std::vector<float> c;
+  ORT_ENFORCE(info.GetAttrs<float>("coefficients", c).IsOK());
+  ORT_ENFORCE(c.size() % targets_ == 0);
+  int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
+  typename Eigen::TensorMap<Eigen::Tensor<float, 2, Eigen::RowMajor, Eigen::DenseIndex>, Eigen::Unaligned> c_tensor(c.data(), targets_, feature_size);
+  //Tranpose the data so that we can use math::MatMul instead of math::Gemm in the compute function
+  Eigen::array<int, 2> shuffle{1, 0};
+  return c_tensor.cast<T>().shuffle(shuffle);
+}
 
-  bool useIntercepts = intercepts_.size() == static_cast<size_t>(targets_);
-  for (int64_t i = 0; i < N; i++)  //for each point
+template <typename T>
+LinearRegressor<T>::LinearRegressor(const OpKernelInfo& info)
+    : OpKernel(info),
+      targets_(GetAttr<int64_t>(info, "targets")),
+      coefficients_(GetCoefficients<T>(info, targets_)),
+      post_transform_(MakeTransform(info.GetAttrOrDefault<std::string>("post_transform", "NONE"))) {
   {
-    std::vector<float> scores;
-    int64_t current_weight_0 = i * stride;
-    for (int j = 0; j < targets_; j++)  //for each target
-    {
-      int64_t current_coeff_0 = j * stride;
-      float weight = 0.f;
-      for (int64_t k = 0; k < stride; k++)  //for each weight
-      {
-        weight = weight + Xdata[current_weight_0 + k] * coefficients_[current_coeff_0 + k];
-      }
-      if (useIntercepts) {
-        weight = weight + intercepts_[j];
-      }
-      scores.push_back(weight);
-    }
-    ::onnxruntime::ml::write_scores(scores, post_transform_, yindex, Y, -1);
-    yindex += scores.size();
+    std::vector<float> c = info.GetAttrsOrDefault<float>("intercepts");
+    intercepts_.resize(c.size());
+    std::copy_n(c.data(), c.size(), intercepts_.data());
+  }
+
+  // A dirty hack to keep the code working as before
+  if (targets_ == 1) {
+    // In RS4, we implemented the PROBIT transform for single class cases,
+    // but the outputted value is most likely NaN
```

Comment:
Why should the ORT code have something that changes the behavior based on things that happened in RS4? That seems like something that should be somewhere other than master.

Reply:
Still output NaN randomly?

Reply:
Just going off the ONNX spec, it seems like this approach is inconsistent as you're ignoring the post transform value completely if targets_ == 1. 
Is that because the spec should say that is the correct behavior, or just because previously we didn't handle this case (or handled it badly)? Or because we're making in implementation choice to not support this usage? If the latter, seems like we should fail or output a warning.

Similarly, just because there wasn't a PROBIT implementation previously for multi class cases, the ONNX spec says it's supported, so also turning off PROBIT if targets_ is != 1 seems like we don't support PROBIT anywhere. If that's the case shouldn't we at least fail the model (it's not going to do what a default user of ORT expects) or at a minimum output a warning message saying we're overriding the post transform value in the model?



---
In reply to: [336695370](https://github.com/microsoft/onnxruntime/pull/1959#discussion_r336695370) [](ancestors = 336695370)

Reply:
But, this is exactly how the old code works. If we change that, it's a breaking change.

Reply:
I changed it to ORT_ENFORCE